pennylane/tracker.py

```diff
@@ -0,0 +1,166 @@
+    executions, but plugins may store additional information with no changes to this class.
+
+    Information is only stored when the class attribute ``tracking`` is set to ``True``. This
+    attribute can be toggled via a context manager and Python's ``with`` statement. Upon entering a
+    context, the stored information is reset, unless ``persistent=True``.
+
+    Args:
+        dev (Device): a PennyLane compatible device
+        callback=None (callable or None): a function of the keywords ``totals``,
+            ``history`` and ``latest``.  Run on each ``record`` call with current values of
+            the corresponding attributes.
+        persistent=False (bool): whether to reset stored information upon
+            entering a runtime context.
+
+
+    **Example**
+
+    Using a ``with`` statement to toggle the tracking mode, we can see the number of executions
+    and shots used to calculate a parameter-shift derivative.
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=1, shots=100)
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        x = np.array(0.1)
+
+        with qml.Tracker(dev) as tracker:
+            qml.grad(circuit)(0.1)
+
+    You can then access the tabulated information through ``totals``, ``history``, and ``latest``:
+
+    >>> tracker.totals
+    {'executions': 3, 'shots': 300}
+    >>> tracker.history
+    {'executions': [1, 1, 1], 'shots': [100, 100, 100]}
+    >>> tracker.latest
+    {'executions': 1, 'shots': 100}
+
+    .. UsageDetails::
+
+    .. note::
+        With backpropagation, this function should take ``qnode.device``
+        instead of the device used to create the QNode.
+
+    Users can pass a custom callback function to the ``callback`` keyword. This
+    function is run each time the ``record()`` method is called. The
+    function passed must accept ``totals``, ``history``, and ``latest`` as
+    keyword arguments:
+
+    >>> def shots_info(totals=dict(), history=dict(), latest=dict()):
+    ...     print("Total shots: ", totals['shots'])
+    >>> with qml.Tracker(circuit.device, callback=shots_info) as tracker:
+    ...     qml.grad(circuit)(0.1)
+    Total shots:  100
+    Total shots:  200
+    Total shots:  300
+
+    By specifying ``persistent=False``, you can reuse the same tracker accross
+    multiple contexts.
+
+    >>> with qml.Tracker(circuit.device, persistent=False) as tracker:
+    ...     circuit(0.1)
+    >>> with tracker:
+    ...     circuit(0.2)
+    >>> tracker.totals['executions']
+    2
+
+    """
+
+    def __init__(self, dev=None, callback=None, persistent=False):
+        self.persistent = persistent
+
+        self.callback = callback
+
+        # same code as self.reset
+        self.totals = dict()
+        self.history = dict()
+        self.latest = dict()
+
+        self.tracking = False
```

Comment:
How about calling this attribute as `on`/`active`? `Tracker.on = False` seems to read slightly better than `Tracker.tracking`